src/compile/kinds/compiler.ts

```diff
@@ -1,14 +1,152 @@
+    CACHE_DIR,
+    getCompilerMDForPlatform,
+    getCompilerPrefixForOs,
+    isSubDir
+} from "./md";
+
+const solc = require("solc");
 
 export abstract class Compiler {
-    constructor(public readonly version: string) {
-        assert(
-            isExact(version),
-            "Version string must contain exact SemVer-formatted version without any operators"
-        );
-    }
+    constructor(public readonly version: string, public readonly path: string) {}
 
     abstract compile(inputJson: SolcInput): Promise<any>;
 }
+
+export class NativeCompiler extends Compiler {
+    async compile(input: SolcInput): Promise<any> {
+        const child = spawn(this.path, ["--standard-json"], {});
+
+        return new Promise((resolve, reject) => {
+            child.stdin.write(JSON.stringify(input), "utf-8");
+            child.stdin.end();
+
+            let stdout = "";
+            let stderr = "";
+
+            child.stdout.on("data", (data) => {
+                stdout += data;
+            });
+
+            child.stderr.on("data", (data) => {
+                stderr += data;
+            });
+
+            child.on("close", (code) => {
+                if (code !== 0) {
+                    reject(`Compiler exited with code ${code}, stderr: ${stderr}`);
+                    return;
+                }
+
+                if (stderr !== "") {
+                    reject(`Compiler exited with non-empty stderr: ${stderr}`);
+                    return;
+                }
+
+                let outJson: any;
+
+                try {
+                    outJson = JSON.parse(stdout);
+                } catch (e) {
+                    reject(e);
+                    return;
+                }
+
+                resolve(outJson);
+            });
+        });
+    }
+}
+
+export class WasmCompiler extends Compiler {
+    async compile(input: SolcInput): Promise<any> {
+        const module = require(this.path);
+        const wrappedModule = solc.setupMethods(module);
+        const output = wrappedModule.compile(JSON.stringify(input));
+
+        return JSON.parse(output);
+    }
+}
+
+type CompilerMapping = [CompilerKind.Native, NativeCompiler] | [CompilerKind.WASM, WasmCompiler];
+
+export async function getCompilerForVersion<T extends CompilerMapping>(
+    version: string,
+    kind: T[0]
+): Promise<T[1] | undefined> {
+    assert(
+        isExact(version),
+        "Version string must contain exact SemVer-formatted version without any operators"
+    );
+
+    let prefix: string | undefined;
+
+    if (kind === CompilerKind.Native) {
+        prefix = getCompilerPrefixForOs();
+    } else if (kind === CompilerKind.WASM) {
+        /**
+         * Using BIN distribution here due to WASM distributions do not have 0.5.17 build and also OOM issues.
+         *
+         * @see https://github.com/ethereum/solidity/issues/10329
+         *
+         * @todo Reconsider this at some point.
+         */
+        prefix = "bin";
```

Comment:
Just FYI, the wasm binary is now available: https://github.com/ethereum/solc-bin/blob/gh-pages/wasm/soljson-v0.5.17+Bcommit.d19bba13.js

Reply:
FYI, OOMs are happening on NodeJS 12 and not happening on NodeJS v16.14.0 (`lts/gallium`). Just in case if you will encounter similar question.